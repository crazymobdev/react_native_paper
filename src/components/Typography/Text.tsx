import * as React from 'react';
import {
  Text as NativeText,
  TextStyle,
  StyleProp,
  StyleSheet,
  I18nManager,
} from 'react-native';
import { useTheme } from '../../core/theming';
import { Font, MD3TypescaleKey, ThemeProp } from '../../types';

type Props = React.ComponentProps<typeof NativeText> & {
  /**
   * @supported Available in v3.x with theme version 3
   *
   * Variant defines appropriate text styles for type role and its size.
   * Available variants:
   *
   *  Display: `display-large`, `display-small`, `display-small`
   *
   *  Headline: `headline-large`, `headline-medium`, `headline-small`
   *
   *  Title: `title-large`, `title-medium`, `title-small`
   *
   *  Label:  `label-large`, `label-medium`, `label-small`
   *
   *  Body: `body-large`, `body-medium`, `body-small`
   */
  variant?: keyof typeof MD3TypescaleKey;
  children: React.ReactNode;
  theme?: ThemeProp;
  style?: StyleProp<TextStyle>;
};

// @component-group Typography

/**
 * Typography component showing styles complied with passed `variant` prop and supported by the type system.
 *
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <>
 *     <Text variant="display-large">Display Large</Text>
 *     <Text variant="display-medium">Display Medium</Text>
 *     <Text variant="display-small">Display small</Text>
 *
 *     <Text variant="headline-large">Headline Large</Text>
 *     <Text variant="headline-medium">Headline Medium</Text>
 *     <Text variant="headline-small">Headline Small</Text>
 *
 *     <Text variant="title-large">Title Large</Text>
 *     <Text variant="title-medium">Title Medium</Text>
 *     <Text variant="title-small">Title Small</Text>
 *
 *     <Text variant="body-large">Body Large</Text>
 *     <Text variant="body-medium">Body Medium</Text>
 *     <Text variant="body-small">Body Small</Text>
 *
 *     <Text variant="label-large">Label Large</Text>
 *     <Text variant="label-medium">Label Medium</Text>
 *     <Text variant="label-small">Label Small</Text>
 *  </>
 * );
 *
 * export default MyComponent;
 * ```
 *
 * @extends Text props https://reactnative.dev/docs/text#props
 */

const Text: React.ForwardRefRenderFunction<{}, Props> = (
  { style, variant, theme: initialTheme, ...rest }: Props,
  ref
) => {
  const root = React.useRef<NativeText | null>(null);
  // FIXME: destructure it in TS 4.6+
  const theme = useTheme(initialTheme);
  const writingDirection = I18nManager.isRTL ? 'rtl' : 'ltr';

  React.useImperativeHandle(ref, () => ({
    setNativeProps: (args: Object) => root.current?.setNativeProps(args),
  }));

  if (theme.isV3 && variant) {
    const stylesByVariant = Object.keys(MD3TypescaleKey).reduce(
      (acc, key) => {
        const { size, weight, lineHeight, tracking } =
          theme.typescale[key as keyof typeof MD3TypescaleKey];

        return {
          ...acc,
          [key]: {
            fontSize: size,
            fontWeight: weight,
            lineHeight: lineHeight,
            letterSpacing: tracking,
            color: theme.colors.onSurface,
          },
        };
      },
      {} as {
        [key in MD3TypescaleKey]: {
          fontSize: number;
          fontWeight: Font['fontWeight'];
          lineHeight: number;
          letterSpacing: number;
        };
      }
    );

    const styleForVariant = stylesByVariant[variant];

    return (
      <NativeText
        ref={root}
        style={[styleForVariant, styles.text, { writingDirection }, style]}
        {...rest}
      />
    );
  } else {
    return (
      <NativeText
        {...rest}
        ref={root}
        style={[
          {
            ...theme.fonts?.regular,
            color: theme.isV3 ? theme.colors?.onSurface : theme.colors.text,
          },
          styles.text,
          style,
        ]}
      />
    );
  }
};

const styles = StyleSheet.create({
  text: {
    textAlign: 'left',
  },
});

export default React.forwardRef(Text);
