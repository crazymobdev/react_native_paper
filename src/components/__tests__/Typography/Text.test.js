import * as React from 'react';
import renderer from 'react-test-renderer';
import Text from '../../Typography/Text.tsx';

it('renders every variant of Text with children as content', () => {
  const content = 'Something rendered as a child content';

  const variants = (
    <>
      <Text variant="display-large">{content}</Text>
      <Text variant="display-medium">{content}</Text>
      <Text variant="display-small">{content}</Text>

      <Text variant="headline-large">{content}</Text>
      <Text variant="headline-medium">{content}</Text>
      <Text variant="headline-small">{content}</Text>

      <Text variant="title-large">{content}</Text>
      <Text variant="title-medium">{content}</Text>
      <Text variant="title-small">{content}</Text>

      <Text variant="body-large">{content}</Text>
      <Text variant="body-medium">{content}</Text>
      <Text variant="body-small">{content}</Text>

      <Text variant="label-large">{content}</Text>
      <Text variant="label-medium">{content}</Text>
      <Text variant="label-small">{content}</Text>
    </>
  );

  const tree = renderer.create(variants).toJSON();

  expect(tree).toMatchSnapshot();
});
