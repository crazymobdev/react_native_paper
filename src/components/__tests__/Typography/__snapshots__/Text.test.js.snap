// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders every variant of Text with children as content 1`] = `
Array [
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Something rendered as a child content
  </Text>,
]
`;
